Sculk Sensor Frequency Registry

Shnupbups opened the pull request:
- Mixin'd to `SculkSensorBlock` to make its `FREQUENCIES` map modifiable.
- Created `SculkSensorFrequencyRegistry` with a dedicated method for registering these frequencies, with all necessary checks to avoid issues, and javadoc to assist in understanding.
- Added tests to the test mod, as well as a block that emits a test event, to ensure it all works.

Unfortunately, the mixin currently uses a Redirect. If this is undesired, feel free to suggest other ways of achieving the end result of being able to modify the frequencies map.

Shnupbups committed:
Sculk Sensor Frequency Registry

- Mixin'd to `SculkSensorBlock` to make its `FREQUENCIES` map modifiable.
- Created `SculkSensorFrequencyRegistry` with a dedicated method for registering these frequencies, with all necessary checks to avoid issues, and javadoc to assist in understanding.
- Added tests to the test mod, as well as a block that emits a test event, to ensure it all works.

Unfortunately, the mixin currently uses a Redirect. If this is undesired, feel free to suggest other ways of achieving the end result of being able to modify the frequencies map.

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/mixin/content/registry/SculkSensorBlockMixin.java:
```diff
@@ -0,0 +1,34 @@
+
+@Mixin(SculkSensorBlock.class)
+public class SculkSensorBlockMixin {
+	@Dynamic("Call to Object2IntMaps.unmodifiable in initialization of FREQUENCIES")
```
Why dynamic?

Shnupbups replied:
It's a decorator annotation provided by mixin specifically for providing context for mixins that aren't immediately obvious where they modify. Similar usage is found in this module's `OxidizableMixin` and `HoneycombMixin` when they inject into lambda bodies. I suppose `<clinit>` isn't altogether hard to imagine, but it's certainly less transparent than a mixin to a visible method, and so I felt it somewhat warranted.

Technici4n replied:
It should only be used for lambda bodies imo, `<clinit>` is quite well known and probably doesn't require the annotation.

Technici4n replied:
This isn't resolved yet since `@Dynamic` is still there? I don't think it makes sense to have it.

Technici4n commented on fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SculkSensorFrequencyRegistry.java:
```diff
@@ -0,0 +1,58 @@
+		int replaced = SculkSensorBlock.FREQUENCIES.getOrDefault(event, 0);
+		SculkSensorBlock.FREQUENCIES.put(event, frequency);
+
+		if (replaced != 0) {
```
Couldn't you check for putInt != 0 in one call?

Shnupbups replied:
Genuinely forgot that the `put` method of maps returns old values if present, good shout. However I'll keep it separate from the `if` due to the variable then being used in the debug log.

Technici4n reviewed (approved):
Looks good to me!

Shnupbups committed:
Update SculkSensorFrequencyRegistry.java

Shnupbups committed:
cleaner javadoc

Juuxel commented on fabric-content-registries-v0/src/main/resources/fabric-content-registries-v0.mixins.json:
```diff
@@ -9,7 +9,8 @@
     "MixinFireBlock",
     "OxidizableMixin",
-    "ShovelItemAccessor"
+    "ShovelItemAccessor",
+    "SculkSensorBlockMixin"
```
```suggestion
    "SculkSensorBlockMixin",
    "ShovelItemAccessor"
```
Alphabetical ordering

Shnupbups committed:
Update fabric-content-registries-v0/src/main/resources/fabric-content-registries-v0.mixins.json

Co-authored-by: Juuxel <6596629+Juuxel@users.noreply.github.com>

Shnupbups committed:
Merge branch '1.19' into 1.19-sculk-frequencies

Shnupbups committed:
Final touches

...get it?

(oh and removed that Dynamic because tech didn't like it)

Shnupbups committed:
Whoops

Forgot to remove the Dynamic import
Also restored the actual contents of the dynamic, as javadoc.

modmuss50 commented:
Looks good, please fix conflicts 👍

Shnupbups committed:
Merge branch '1.19' into 1.19-sculk-frequencies
